peer/peerlist/list.go

```diff
@@ -0,0 +1,588 @@
+// Must be run in a mutex.Lock()
+func (pl *List) removeAllUnavailablePeers(toRemove map[string]*peerThunk) []*peerThunk {
+	thunks := make([]*peerThunk, 0, len(toRemove))
+	for id, t := range toRemove {
+		thunks = append(thunks, t)
+		delete(toRemove, id)
+	}
+	return thunks
+}
+
+// releaseAll will iterate through a list of peers and call release
+// on the transport
+func (pl *List) releaseAll(errs error, peers []*peerThunk) error {
+	for _, t := range peers {
+		if err := pl.transport.ReleasePeer(t.peer, t); err != nil {
+			errs = multierr.Append(errs, err)
+		}
+	}
+	return errs
+}
+
+// removePeerIdentifier will go remove references to the peer identifier and release
+// it from the transport
+// Must be run in a mutex.Lock()
+func (pl *List) removePeerIdentifier(pid peer.Identifier) error {
+	t, err := pl.removePeerIdentifierReferences(pid)
+	if err != nil {
+		// The peer has already been removed
+		return err
+	}
+
+	return pl.transport.ReleasePeer(pid, t)
+}
+
+// removePeerIdentifierReferences will search through the Available and Unavailable Peers
+// for the PeerID and remove it
+// Must be run in a mutex.Lock()
+func (pl *List) removePeerIdentifierReferences(pid peer.Identifier) (*peerThunk, error) {
+	if t := pl.availablePeers[pid.Identifier()]; t != nil {
+		return t, pl.removeFromAvailablePeers(t)
+	}
+
+	if t, ok := pl.unavailablePeers[pid.Identifier()]; ok && t != nil {
+		pl.removeFromUnavailablePeers(t)
+		return t, nil
+	}
+
+	return nil, peer.ErrPeerRemoveNotInList(pid.Identifier())
+}
+
+// removeFromAvailablePeers remove a peer from the Available Peers list the
+// Peer should already be validated as non-nil and in the Available list.
+// Must be run in a mutex.Lock()
+func (pl *List) removeFromAvailablePeers(t *peerThunk) error {
+	delete(pl.availablePeers, t.peer.Identifier())
+	pl.availableChooser.Remove(t, t.Subscriber())
+	t.SetSubscriber(nil)
+	return nil
+}
+
+// removeFromUnavailablePeers remove a peer from the Unavailable Peers list the
+// Peer should already be validated as non-nil and in the Unavailable list.
+// Must be run in a mutex.Lock()
+func (pl *List) removeFromUnavailablePeers(t *peerThunk) {
+	delete(pl.unavailablePeers, t.peer.Identifier())
+}
+
+// Choose selects the next available peer in the peer list
+func (pl *List) Choose(ctx context.Context, req *transport.Request) (peer.Peer, func(error), error) {
+	if err := pl.once.WaitUntilRunning(ctx); err != nil {
+		return nil, nil, pl.newNotRunningError(err)
+	}
+
+	for {
+		p, err := pl.choose(ctx, req)
+		if p != nil {
+			t := p.(*peerThunk)
+			pl.notifyPeerAvailable()
+			t.onStart()
+			return t.peer, t.boundOnFinish, nil
+		}
+		// This is possible only if the available peers chooser is empty at the
+		// time of the call.
+		// We race again each time we add a peer to the available peers.
+		// Since we call Choose outside a lock, it is always possible that it's
+		// empty.
+		if !yarpcerrors.IsUnavailable(err) {
+			return nil, nil, err
```

Comment:
I think we need to notifyPeerAvailable here, otherwise we have an opportunity here for infinite blocking on a peer list.

Given the following sequence:
- 5 requests queued while no peers were available
- 1 peer is added (peerAvailableEvent has element)
- first request in the queue is unblocked on the peer list
- first request returns an error (not unavailable)
- request hits this line and exits the peer chooser
- 4 other requests are still blocked on the peerAvailableEvent, these requests will timeout or wait until another peer event/next successful request (normally this won't apply, this was just our solution to unblocking when there are a queue of requests)

This is definitely and edge case, but we should make sure that available choosers respect the contract to not return Unavailable errors unless they have no peers to choose from (as opposed to not having a peer they find appropriate).